### PR TITLE
[CORE-538] Apply dYdX patches to Cosmos SDK 0.50.1 fork

### PR DESCRIPTION
**Pull request by lcwik**
## Description

Apply dYdX patches to Cosmos SDK 0.50.1 fork

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: baseapp/abci.go [hunk 1]
```diff
@@ -38,6 +38,9 @@ const (
 )
 
 func (app *BaseApp) InitChain(req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
+	app.mtx.Lock()
+	defer app.mtx.Unlock()
+
 	if req.ChainId != app.chainID {
 		return nil, fmt.Errorf("invalid chain-id on InitChain; expected: %s, got: %s", app.chainID, req.ChainId)
 	}
```
Comment (teddyding) [inline baseapp/abci.go:46]: For my understanding, why did we leave out these [lines](https://github.com/dydxprotocol/cosmos-sdk/blob/bdf96fddf421e90574e53af383c1468fdb33d7a3/baseapp/abci.go#L74-L82) (back-ported [here](https://github.com/dydxprotocol/cosmos-sdk/commit/ae5ce197934ccb21acd6dba43d875ecb77c27b2f))?

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -184,6 +186,9 @@ type BaseApp struct {
 	// including the goroutine handling.This is experimental and must be enabled
 	// by developers.
 	optimisticExec *oe.OptimisticExecution
+
+	// Used to synchronize the application when using an unsynchronized ABCI++ client.
+	mtx sync.Mutex
```
Comment (teddyding) [inline baseapp/baseapp.go:191]: Question on `DeliverTx`:

I see that on `v0.47` fork we take the `app.mtx` lock https://github.com/dydxprotocol/cosmos-sdk/blob/bdf96fddf421e90574e53af383c1468fdb33d7a3/baseapp/abci.go#L418-L429 in `DeliverTx`. How has this changed in `v0.50`? This is the corresponding `deliverTx` [function](https://github.com/dydxprotocol/cosmos-sdk/blob/dccb1f13c3560bac441c11f93151bea876649bb1/baseapp/baseapp.go#L729) - is it intended to not take the lock?

File: baseapp/abci.go [hunk 3]
```diff
@@ -901,6 +937,9 @@ func (app *BaseApp) checkHalt(height int64, time time.Time) error {
 // against that height and gracefully halt if it matches the latest committed
 // height.
 func (app *BaseApp) Commit() (*abci.ResponseCommit, error) {
+	app.mtx.Lock()
+	defer app.mtx.Unlock()
+
 	header := app.finalizeBlockState.ctx.BlockHeader()
 	retainHeight := app.GetBlockRetentionHeight(header.Height)
```
Comment (teddyding) [inline baseapp/abci.go:945]: Question: do we still need the `commiter` logic from [here](https://github.com/dydxprotocol/cosmos-sdk/blame/bdf96fddf421e90574e53af383c1468fdb33d7a3/baseapp/abci.go#L512-L514)? Or is that replaced by the `prepareCheckStater` below?

File: baseapp/baseapp.go [hunk 4]
```diff
@@ -784,6 +789,145 @@ func (app *BaseApp) endBlock(ctx context.Context) (sdk.EndBlock, error) {
...
+		Events:       sdk.EmptyEvents().ToABCIEvents(),
+		MsgResponses: msgResponses,
+	}
+
+	return gInfo, result, anteEvents, err
+}
+
 // runTx processes a transaction within a given execution mode, encoded transaction
 // bytes, and the decoded transaction itself. All state transitions occur through
 // a cached Context depending on the mode provided. State only gets persisted
```
Comment (teddyding) [inline baseapp/baseapp.go:933]: Question for `runTx()` below: do we still need these [checks](https://github.com/dydxprotocol/cosmos-sdk/blob/bdf96fddf421e90574e53af383c1468fdb33d7a3/baseapp/baseapp.go#L763-L765)?

**Comment by teddyding** [comment]
To confirm, do we still need to port over these [changes](https://github.com/dydxprotocol/cosmos-sdk/compare/e13fdea30f46ea1e420e666de41b39b3b2978914...bdf96fddf421e90574e53af383c1468fdb33d7a3) from the `v0.47` fork?

**Comment by teddyding** [comment]
For my understanding, how do we plan to test the upgrade from `v0.47` fork to `v0.50` fork? Involves a lot of changes all over the places, and it seems easy to miss something to port over.

**Comment by lcwik** [comment]
> To confirm, do we still need to port over these [changes](https://github.com/dydxprotocol/cosmos-sdk/compare/e13fdea30f46ea1e420e666de41b39b3b2978914...bdf96fddf421e90574e53af383c1468fdb33d7a3) from the `v0.47` fork?

Just added these now to this PR.

**Comment by lcwik** [comment]
> For my understanding, how do we plan to test the upgrade from `v0.47` fork to `v0.50` fork? Involves a lot of changes all over the places, and it seems easy to miss something to port over.

This is TBD, my thoughts around this are:
* ensure existing unit tests pass
* deploy to `dev` to ensure it continues to run
* perform upgrade testing (either using the upgrade test framework if it exists or performing upgrades in a persistent environment (dev/staging/testnet))

File: baseapp/baseapp.go (hunk 4 above)
Comment (lcwik) [reply baseapp/baseapp.go:933]: Yes, added it back and updated the `[STAB-19] Reduce the size of the critical section` commit to contain it to keep the history clean.

File: baseapp/abci.go (hunk 3 above)
Comment (lcwik) [reply baseapp/abci.go:945]: It is replaced by `prepareCheckStater` which @prettymuchbryce pushed upstream as the replacement for the `committer` logic.

File: baseapp/baseapp.go (hunk 2 above)
Comment (lcwik) [reply baseapp/baseapp.go:191]: Cosmos was re-entrant when delivering txs during block commit which is why we needed the ability to selectively take the lock. CometBFT pushed down block commit logic of individually calling `DeliverTx` into `FinalizeBlock` which means that we could elevate the need for acquiring the lock to `FinalizeBlock` removing it from `deliverTx` which also meant that we didn't care any more that Cosmos was re-entrant for this method or not.

**Comment by lcwik** [comment]
I was able to add the last of the remaining commits.

I flattened these two commits together (https://github.com/dydxprotocol/cosmos-sdk/commit/540205509c24f79581cea6d08c7226eeb31785f5, https://github.com/dydxprotocol/cosmos-sdk/commit/b95c66dedbd525a1b3d0f6b851d4fd351887fa23).

And the CLOB-930 commit was able to be simplified a bunch because of changes that were released in 0.50 (https://github.com/cosmos/cosmos-sdk/pull/15041).

File: baseapp/baseapp.go (hunk 2 above)
Comment (teddyding) [reply baseapp/baseapp.go:191]: Thanks for the explanation. I see from [here](https://github.com/cometbft/cometbft/blob/v0.38.0/CHANGELOG.md#v0380) tha `BeginBlock`, `DeliverTx`, and `EndBlock` are coalesced into `FinalizeBlock`

**Review by teddyding — APPROVED** [review]

**Pull request merged**